Mathlib/Topology/Semicontinuous.lean

```diff
@@ -295,6 +352,40 @@ theorem Continuous.lowerSemicontinuous {f : α → γ} (h : Continuous f) : Lowe
   fun _x => h.continuousAt.lowerSemicontinuousAt
 #align continuous.lower_semicontinuous Continuous.lowerSemicontinuous
 
+
+end
+
+section
+
+variable {γ : Type*} [CompleteLinearOrder γ] [DenselyOrdered γ]
+
+variable [TopologicalSpace γ] [FirstCountableTopology α] [FirstCountableTopology γ] [OrderTopology γ]
+
+theorem lowerSemicontinuous_iff_IsClosed_epigraph {f : α → γ} :
+    LowerSemicontinuous f ↔ IsClosed {p : α × γ | f p.1 ≤ p.2} := by
+  constructor
+  · simp only [lowerSemicontinuous_iff_le_liminf]
+    intro hf; apply IsSeqClosed.isClosed
+    intro f' ⟨x', y'⟩ hxy cxy
+    rw [Prod.tendsto_iff] at cxy
+    let x : ℕ -> α := fun (n : ℕ) => (f' n).1
+    calc
+      f x' ≤ liminf f (𝓝 x') := hf x'
+      _ ≤ liminf (f ∘ x) atTop := by
+        simp only [liminf_eq, liminf_eq]
+        exact sSup_le_sSup (fun _ fa => (eventually_iff_seq_eventually.mp fa) x cxy.1)
+      _ ≤ liminf (fun (n : ℕ) => (f' n).2) atTop :=
+        liminf_le_liminf (eventually_of_forall (fun n => by convert hxy n))
+      _ = y' := (cxy.2).liminf_eq
+  simp only [lowerSemicontinuous_iff_isClosed_preimage]
+  intro hf y; apply IsSeqClosed.isClosed
+  exact fun _ _ xns cx =>
+    IsClosed.isSeqClosed hf (fun n => xns n) (Tendsto.prod_mk_nhds cx tendsto_const_nhds)
```

Comment:
I would like to do this proof without using `IsSeqClosed`, because I believe it's true generally, not just for sequential spaces.
The backwards direction is easy to adapt:
```suggestion
  intro hf y
  exact hf.preimage (Continuous.Prod.mk_left y)
```
For the forwards direction I'm quite sure it's still true. But I want a lemma that `s` is closed iff `s` is closed under limits by an arbitrary filter. I believe this is true in an arbitrary topological space. And then we can mimick the proof done e.g. here: https://math.stackexchange.com/questions/2276334/lower-semicontinuity-and-closedness-equivalence
However, I cannot find this in the library (it's kinda the reverse of `IsClosed.mem_of_tendsto`).

@PatrickMassot, maybe you can help?

Reply:
Sure. Indeed it seems the relevant characterization of closed sets is mysteriously absent from Mathlib, although the very closely related `isClosed_iff_clusterPt` is there. You can put
```lean
lemma isClosed_iff_forall_filter {X : Type*} [TopologicalSpace X] {s : Set X} :
    IsClosed s ↔ ∀ x, ∀ F : Filter X, F.NeBot → F ≤ 𝓟 s → F ≤ 𝓝 x → x ∈ s := by
  rw [isClosed_iff_clusterPt]
  exact ⟨fun hs x F F_ne FS Fx ↦ hs _ <| NeBot.mono F_ne (le_inf Fx FS),
         fun hs x hx ↦ hs x (𝓝 x ⊓ 𝓟 s) hx inf_le_right inf_le_left⟩
```
somewhere close (he he) to `isClosed_iff_clusterPt`. The `F ≤ 𝓟 s → F ≤ 𝓝 x` looks a bit silly, but somehow it felt easier to read than the equivalent `F ≤ 𝓟 s ⊓ 𝓝 x` which would make the proof slightly shorter. The silly version also saves some unpacking or repacking when using the lemma.

With this lemma at hand, the theorem under discussion becomes
```lean
theorem lowerSemicontinuous_iff_IsClosed_epigraph {f : α → γ} :
    LowerSemicontinuous f ↔ IsClosed {p : α × γ | f p.1 ≤ p.2} := by
  constructor
  · rw [lowerSemicontinuous_iff_le_liminf, isClosed_iff_forall_filter]
    rintro hf ⟨x, y⟩ F F_ne h h'
    rw [nhds_prod_eq, le_prod] at h'
    calc f x ≤ liminf f (𝓝 x) := hf x
    _ ≤ liminf f (map Prod.fst F) := liminf_le_liminf_of_le h'.1
    _ ≤ liminf Prod.snd F := liminf_le_liminf <| (eventually_principal.2 fun x ↦ id).filter_mono h
    _ = y := h'.2.liminf_eq    
  · rw [lowerSemicontinuous_iff_isClosed_preimage]
    exact fun hf y ↦ hf.preimage (Continuous.Prod.mk_left y)
```
Of course one shouldn't forget to remove the useless assumption on the topology of `γ`.

Final note: before realizing they were true by definition, I thought I would need
```lean
lemma Filter.liminf_comp {α β γ : Type*} [ConditionallyCompleteLattice α] (u : β → α) (v : γ → β)
    (f : Filter γ) : liminf (u ∘ v) f = liminf u (map v f) := rfl

lemma Filter.limsup_comp {α β γ : Type*} [ConditionallyCompleteLattice α] (u : β → α) (v : γ → β)
    (f : Filter γ) : limsup (u ∘ v) f = limsup u (map v f) := rfl
```
I think those lemmas should be somewhere near the definition of limsup and liminf since otherwise people may look for them just as I did.

Reply:
Thanks Patrick, I PR'd your small lemmas in #8898, so this PR can use them after that PR is merged.

Reply:
#8898 is merged, so this PR can be rebased on master.